ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -598,45 +600,67 @@ def get_physical_pages(self, for_fileIds=None, for_pageIds=None, return_divs=Fal
-                            any([isinstance(p, typing.Pattern) and p.fullmatch(page_id) for p in pageId_patterns]):
-                        if return_divs:
-                            ret.append(self._page_cache[page_id])
-                        else:
-                            ret.append(page_id)
-            else:
-                for page in self._tree.getroot().xpath(
-                        'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"]',
-                        namespaces=NS):
-                    page_id = page.get('ID')
-                    if page_id in pageId_patterns or \
-                            any([isinstance(p, typing.Pattern) and p.fullmatch(page_id) for p in pageId_patterns]):
-                        if return_divs:
-                            ret.append(page)
-                        else:
-                            ret.append(page_id)
+                    page_attr_patterns += [pageId_token]
+            if page_attr_patterns:
+                if self._cache_flag:
+                    # determine attr to look for before iterating
+                    try:
+                        attr = next(a for a in METS_PAGE_DIV_ATTRIBUTE if (
+                                    any(p in self._page_cache[a] for p in page_attr_patterns) or \
+                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) \
+                                        for p in page_attr_patterns \
+                                        for attr_val in self._page_cache[a]]
+                                    )))
+                        for attr_val in self._page_cache[attr].keys():
+                            if attr_val in page_attr_patterns or \
+                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) for p in page_attr_patterns]):
+                                if return_divs:
+                                    ret.append(self._page_cache[attr][attr_val])
+                                else:
+                                    ret.append(attr_val)
+                    except StopIteration:
+                        log.debug(f"No pattern matches any keys of any of the _page_caches. patterns: {page_attr_patterns}")
+                else:
+                    # determine attr during iterating
+                    attr = None
+                    for page in self._tree.getroot().xpath(
+                            'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"]',
+                            namespaces=NS):
+                        try:
+                            if not attr:
+                                attr = next(a for a in METS_PAGE_DIV_ATTRIBUTE if \
+                                    page.get(a.name) in page_attr_patterns or \
+                                    any([isinstance(p, typing.Pattern) and p.fullmatch(page.get(a.name)) for p in page_attr_patterns]))
```

Comment:
Here it's easier to achieve, because there can be an efficient single loop across all pages (regardless of type), so the facets would be the inner loop (but still deleting matches from `page_attr_patterns` for efficiency).